### PR TITLE
GEOPY-1514: Add EM survey types to object dropdown in octree-creation-app ui.json

### PR DESCRIPTION
**Pull request by benk-mira**
**GEOPY-1514 - Add EM survey types to object dropdown in octree-creation-app ui.json**
…ontains a complement.

### CODE REVIEW

File: octree_creation_app/driver.py [hunk 1]
```diff
@@ -72,50 +72,82 @@ def octree_from_params(params: OctreeParams):
...
-                mesh = OctreeDriver.refine_tree_from_surface(
-                    mesh,
-                    refinement_object,
-                    levels,
-                    params.diagonal_balance,
-                    max_distance=getattr(params, value["distance"]),
-                )
+        print("Finalizing . . .")
+        mesh.finalize()
+        octree = treemesh_2_octree(params.geoh5, mesh, name=params.ga_group_name)
```
Comment (domfournier) [inline octree_creation_app/driver.py:95]: Would you mind adding a logging about being done?

**Review by domfournier — COMMENTED** [review]
Looks great.

File: octree_creation_app/driver.py (hunk 1 above)
Comment (benk-mira) [reply octree_creation_app/driver.py:95]: done.

**Review by domfournier — APPROVED** [review]

File: octree_creation_app-assets/uijson/octree_mesh.ui.json [hunk 2]
```diff
@@ -26,7 +26,10 @@
             "{b99bd6e5-4fe1-45a5-bd2f-75fc31f91b38}",
             "{0b639533-f35b-44d8-92a8-f70ecff3fd26}",
             "{9b08bb5a-300c-48fe-9007-d206f971ea92}",
-            "{19730589-fd28-4649-9de0-ad47249d9aba}"
+            "{19730589-fd28-4649-9de0-ad47249d9aba}",
+            "{b3a47539-0301-4b27-922e-1dde9d882c60}",
+            "{a81c6b0a-f290-4bc8-b72d-60e59964bfe8}",
+            "{41018a45-01a0-4c61-a7cb-9f32d8159df4}"
```
Comment (domfournier) [inline octree_creation_app-assets/uijson/octree_mesh.ui.json:32]: Missing a couple here:
```suggestion
            "{41018a45-01a0-4c61-a7cb-9f32d8159df4}",
            "{deebe11a-b57b-4a03-99d6-8f27b25eb2a8}",
            "{275ecee9-9c24-4378-bf94-65f3c5fbe163}"
```

File: octree_creation_app/driver.py [hunk 3]
```diff
@@ -72,50 +72,82 @@ def octree_from_params(params: OctreeParams):
...
-        mesh.finalize()
-        octree = treemesh_2_octree(params.geoh5, mesh, name=params.ga_group_name)
+        else:
+            raise NotImplementedError(
+                f"Refinement for object {type(refinement_object)} is not implemented."
+            )
 
-        return octree
+        if hasattr(refinement_object, "complement"):
+            mesh = OctreeDriver.refine_by_object_type(
```
Comment (domfournier) [inline octree_creation_app/driver.py:141]: Unfortunately, this sends it into an infinite loop. You will need an argument to only execute the complement once.

Comment (benk-mira) [reply octree_creation_app/driver.py]: using a loop now instead

**Review by domfournier — DISMISSED** [review]

**Review by domfournier — DISMISSED** [review]
I did the change to refinements

**Review by andrewg-mira — DISMISSED** [review]

**Review by domfournier — APPROVED** [review]

**Pull request merged**